gpm_api/tests/io/test_checks.py

```diff
@@ -0,0 +1,497 @@
+
+
+def test_check_product_level(
+    product_levels: List[str],
+) -> None:
+    """Test check_product_level()"""
+
+    # Test types that aren't strings
+    for product_level in [123, None]:
+        with pytest.raises(ValueError):
+            checks.check_product_level(product_level)
+
+    # Test a product_level that does exist
+    for product_level in product_levels:
+        res = checks.check_product_level(product_level)
+        assert (
+            res is None
+        ), f"Function returned {res} for product_level {product_level}, expected None"
+
+    # Test a product_level that doesn't exist
+    for product_level in ["NOT", "A", "LEVEL"]:
+        with pytest.raises(ValueError):
+            checks.check_product_level(product_level)
+
+
+def test_check_product_validity(
+    product_types: List[str],
+) -> None:
+    """Test check_product_validity()"""
+
+    # Test a product that does exist
+    for product_type in product_types:
+        for product in available_products(product_type=product_type):
+            res = checks.check_product_validity(product, product_type=product_type)
+            assert res is None, f"Function returned {res} for product {product}, expected None"
+
+    # Test a product that doesn't exist
+    for product_type in product_types:
+        for product in [("IMERG"), 123, None]:
+            with pytest.raises(ValueError):
+                checks.check_product_validity(product, product_type=product_type)
+            # Test a None product type
+            with pytest.raises(ValueError):
+                checks.check_product_validity(product, product_type=None)
+
+
+def test_check_time() -> None:
+    """Test that time is returned a datetime object from varying inputs"""
+
+    # Test a string
+    res = checks.check_time("2014-12-31")
+    assert isinstance(res, datetime.datetime)
+    assert res == datetime.datetime(2014, 12, 31)
+
+    # Test a string with hh/mm/ss
+    res = checks.check_time("2014-12-31 12:30:30")
+    assert isinstance(res, datetime.datetime)
+    assert res == datetime.datetime(2014, 12, 31, 12, 30, 30)
+
+    # Test a string with <date>T<time>
+    res = checks.check_time("2014-12-31T12:30:30")
+    assert isinstance(res, datetime.datetime)
+    assert res == datetime.datetime(2014, 12, 31, 12, 30, 30)
+
+    # Test a datetime object
+    res = checks.check_time(datetime.datetime(2014, 12, 31))
+    assert isinstance(res, datetime.datetime)
+    assert res == datetime.datetime(2014, 12, 31)
+
+    # Test a datetime timestamp with h/m/s/ms
+    res = checks.check_time(datetime.datetime(2014, 12, 31, 12, 30, 30, 300))
+    assert isinstance(res, datetime.datetime)
+    assert res == datetime.datetime(2014, 12, 31, 12, 30, 30, 300)
+
+    # Test a np.datetime64 object of "datetime64[s]"
+    res = checks.check_time(np.datetime64("2014-12-31"))
+    assert isinstance(res, datetime.datetime)
+    assert res == datetime.datetime(2014, 12, 31)
+
```

Comment:
Let's also test a numpy array with inside a datetime object and a `pd.Timestamp`
`np.array([datetime.datetime(2014, 12, 31, 12, 30, 30)])`

Reply:
I have made the test cases for this